### ISSUE TITLE
cypress doesn't seem to wait for text to appear on the page

### ISSUE THREAD

**Issue by x-yuri** [cypress-io/cypress#3356]
### Current behavior:

I occasionally get 400s in Developer Tools from a Django app. In the shell Django says:

> The request's session was deleted before the request completed. The user may have logged out in a concurrent request, for example.

That happens because the next HTTP request is sent before the previous one has finished:

```
/order: load session (2n40w0tro8okxc08y5ab5pzhepva7iy7)
/login: load session (2n40w0tro8okxc08y5ab5pzhepva7iy7)
/login: save session (c1i4iaggq68jjn56u9u4ptsqt8fmguc2)
/order: save session (2n40w0tro8okxc08y5ab5pzhepva7iy7)
```

The test code:

```js
cy.get('.button').click();   // send order form
// it's sent using AJAX
// once XHR finishes, it replaces the form with thank you message
cy.get('.page').should('contain', 'Thanks');
// it's like it doesn't wait for "Thanks" to appear on the page
cy.request('POST', '/login');
```

Now that I look into `cypress`' source code, `request` [command][1] immediately follows `get` command. Like `should` is not chained. Although I'm not sure. Is there a way to enable javascript logging on the client side (Developer tools)?

[1]: https://github.com/cypress-io/cypress/blob/v3.1.5/packages/driver/src/cypress/cy.coffee#L206

### Desired behavior:

no 400s

### Steps to reproduce: (app code and test code)

How do I go about it?... I'm not sure if I create a simple Django app, I'll be able to reproduce the issue. It doesn't reproduce every time even with the app I'm working on.

Could you possibly point me at the relevant parts of the code, so that I could debug and find the reason myself?

### Versions

`cypress-3.1.4`, Arch Linux

**Comment by jennifer-shehane** [comment]
Could you post a screenshot of the test failing in Cypress - specifically of the Command Log and the error it prints?

Our docs have a section on [Troubleshooting Cypress](https://on.cypress.io/debugging#Troubleshooting-Cypress) that you may find helpful, particularly logging of the events.

**Comment by x-yuri** [comment]
No tests fail. There's only an error in Developer Tools. Specifically this one:

![](https://i.imgur.com/12BWpru.png)

That happens because two requests get handled in parallel. One of them (login) recreates a session. The other one notices the fact, and complains (suspicious operation).

Let me describe it in more detail. Consider an online shop with an order form. User clicks "Place an Order." An AJAX request is sent, that creates an order in database. On response client-side code replaces the form with a "Thank you" message.

The test clicks "Place an Order." Waits for the "Thank you" message. And supposedly then sends the login request. So, there are 2 HTTP requests involved here. And every so often it just so happens that login request starts after submit one, but finishes before it.

Let me put it this way. There are 3 points each request passes that are of concern here: 1) [loading][1] session, 2) [saving][2] session, 3) sending [reply][3]. When Django starts sending the reply, status of the request on the Network tab of Developer Tools changes from `(pending)` to `200 OK`. When it reaches the stated `Content-Length` `XMLHTTPRequest`'s `load` event is triggered (`XMLHTTPRequest.readyState == 4 (DONE)`).

[1]: https://github.com/django/django/blob/2.1.5/django/contrib/sessions/backends/base.py#L195
[2]: https://github.com/django/django/blob/2.1.5/django/contrib/sessions/backends/db.py#L87
[3]: https://github.com/python/cpython/blob/v3.7.2/Lib/wsgiref/handlers.py#L180

In other words, what happens is:

    /order: load session
    /login: load session
    /login: save session
    /order: save session
    /order: complain about new session

Now regarding the client side, there are 2 important points here: 1) after it was confirmed that the page contains the "Thank you" message, 2) before login request is sent.

[4]: https://github.com/cypress-io/cypress/blob/v3.1.5/packages/driver/src/cy/chai.coffee#L146-L151
[5]: https://github.com/cypress-io/cypress/blob/v3.1.5/packages/driver/src/cy/commands/request.coffee#L189

And here the order of events is as follows:

    the page contains "Thank you"
    send login request
    got 400 from /order request

I can't make head or tails out of it, can you? I'm going to try and compile a test case...

**Comment by acyuta108** [comment]
@x-yuri I'm not sure your problem is a Cypress issue unless you only witness the error on the Cypress test runs and not local development (manual) test or in production. It looks to me like more of an async issue in your code.
Now, to ensure AJAX code is working properly in Cypress tests, I always write my tests according to the [ Network Requests guide in the docs](https://docs.cypress.io/guides/guides/network-requests.html#Waiting)
I recommend you aliasing the request  and wait for it to complete before testing the contains 'Thank you'. Something like this:
```
cy.server()
cy.route('POST', '/order').as(orderSent)
cy.get('.button').click();   // send order form
cy.wait('@orderSent')
cy.get('.page').should('contain', 'Thanks');
```
I hope this helps.

**Comment by x-yuri** [comment]
Okay, I've figure it out. Here's a [repo][1] to reproduce the issue (branch `cypress-should-contain`). But well, that's probably not a bug. Although it was totally unexpected.

It didn't occur to me that `contain` can match text that is not visible. And content of script tags for that matter. The "Thank you" message was on the page right from the start. All the script had to do is make it visible. So, `should('contain', 'Thank you')` matched right away before the message had a chance to become visible. In other words, before the first request finished. And following that the second request would start. That is, the two requests start in parallel.

So, I think I'm closing it. Sorry for bothering. It might make sense to note this somewhere in the docs, if not already there.

[1]: https://github.com/x-yuri/issues/tree/cypress-should-contain
